program-runtime/src/loaded_programs.rs

```diff
@@ -895,24 +934,62 @@ impl<FG: ForkGraph> LoadedPrograms<FG> {
         self.unload_program_entries(sorted_candidates.iter().take(num_to_unload));
     }
 
+    /// Evicts programs using 2's random selection, choosing the least used program out of the two entries.
+    /// The eviction is performed enough number of times to reduce the cache usage to the given percentage.
+    pub fn evict_using_2s_random_selection(&mut self, shrink_to: PercentageInteger, now: Slot) {
+        let mut candidates = self.get_flattened_entries();
+        let num_to_unload = candidates
+            .len()
+            .saturating_sub(shrink_to.apply_to(MAX_LOADED_ENTRY_COUNT));
+        fn random_index_and_usage_counter(
+            candidates: &[(Pubkey, Arc<LoadedProgram>)],
+            now: Slot,
+        ) -> (usize, u64) {
+            let mut rng = thread_rng();
+            let index = rng.gen_range(0..candidates.len());
+            let usage_counter = candidates
+                .get(index)
+                .expect("Failed to get cached entry")
+                .1
+                .decayed_usage_counter(now);
+            (index, usage_counter)
+        }
+
+        for _ in 0..num_to_unload {
+            let (index1, usage_counter1) = random_index_and_usage_counter(&candidates, now);
+            let (index2, usage_counter2) = random_index_and_usage_counter(&candidates, now);
+
+            let (program, entry) = if usage_counter1 < usage_counter2 {
+                candidates.remove(index1)
```

Comment:
We don't care about the order of candidates here right? If so we could make
these swap_remove and get O(1) instead of O(n) because of the memmoves